Udp эхо-сервер

tritseratop opened the pull request:
Добавлены udp-клиент и udp-сервер

Commit:
Добавлен udp клиент и сервер (не до конца)

Commit:
feat(udp-сервер): реализована логика udp-сервера

Commit:
chore(ConnectionSettings): изменен регистр параметра protocol

Commit:
fix(udp-клиент): добавлен endpoint сервера, изменен прием сообщения

bachmak commented on CMakeLists.txt:
```diff
@@ -6,10 +6,14 @@ include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
 conan_basic_setup()
 
 aux_source_directory(client CLIENT_SOURCES)
```
вместо дублирования вызовов aux_source_directory можно сформировать список директорий, а aux_source_directory вызывать в цикле, итерируясь по этому списку

bachmak commented on client/main.cpp:
```diff
@@ -1,14 +1,26 @@
+  auto [protocol, host_name, host_port] =
+      ese::ReadConnectionSettings(argc, argv);
 
+  auto client = [&]() -> std::unique_ptr<ese::Client> {
```
Лучше вынести эту логику в отдельный класс (или функцию) фабрики

tritseratop replied:
вынесла в функцию, но не знаю куда вынести функцию

bachmak commented on client/tcp/client.h:
```diff
@@ -0,0 +1,34 @@
+  explicit Client(Context& context, const Ip& host, Port host_port,
+                  Logger& logger);
+
+ private:
```
Этот метод следует сделать публичным, т.к. обращение к этому классу может осуществляться не только через базовый класс

bachmak commented on client/udp/client.h:
```diff
@@ -0,0 +1,33 @@
+
+ private:
+  Socket socket_;
+  Endpoint endpoint_;
```
Предлагаю сделать type alias на этот тип в forwards.h

bachmak commented on forwards.h:
```diff
@@ -5,14 +5,22 @@
+class Logger;
+
+using Context = boost::asio::io_context;
 using Buffer = boost::asio::streambuf;
```
Использовать один тип буфера везде (либо boost::asio::streambuf, либо std::array<char, 1024>, либо что-то еще)

tritseratop replied:
убрала использование std::array<char, 1024>, но там все равно еще используется boost::asio::buffer помимо boost::asio::streambuf

bachmak commented on server/tcp/server.cpp:
```diff
@@ -1,20 +1,21 @@
-      [this](ErrorCode ec, Socket sock) { OnAccepted(ec, std::move(sock)); });
+void Server::Start() {
+  logger_.LogLine("waiting for connections...");
+  auto on_accepted = [this](ErrorCode ec, Socket sock) {
```
Если здесь вынесли лямбду в отдельную переменную, стоит поддерживать консистентность с аналогичными местами

bachmak commented on server/udp/server.cpp:
```diff
@@ -0,0 +1,50 @@
+  if (ec) {
+    ESE_LOG_EC(logger_, ec)
+    return;
+  }
```
\n

bachmak reviewed (commented):
Есть небольшие недочеты, прошу исправить

Commit:
refactor(CMakeLists): Добавлена итерация по директориям

Вместо дублирования вызовов aux_source_directory сформирован список директорий, а aux_source_directory вызывается в цикле по этому списку

Commit:
style(udp- и tcp-сервер): Форматирование

Commit:
style(udp- и tcp-клиент): Исправлена видимость метода, удален буфер

private метод Start исправлен на public, удален буфер std::array<char, 1024>

Commit:
refactor(main клиент и сервер): Добавлена функция-фабрика

Commit:
refactor(CreateServer, CreateClient): перенос функции-фабрики

bachmak commented on CMakeLists.txt:
```diff
@@ -27,6 +27,8 @@ foreach(SOURCE_DIR ${SERVER_SOURCE_DIRS})
     aux_source_directory(${SOURCE_DIR} SERVER_SOURCES)
 endforeach()
 
+message(STATUS ${SERVER_SOURCES})
```
Видимо, этот message отладочный - если так, стоит его удалить

Commit:
chore(CMakeLists): удалена функция message

bachmak reviewed (commented):
Принято
